source/s3_meta_request.c

```diff
@@ -1588,7 +1596,38 @@ struct aws_future_bool *aws_s3_meta_request_read_body(
     AWS_PRECONDITION(meta_request);
     AWS_PRECONDITION(buffer);
 
-    return aws_async_input_stream_read_to_fill(meta_request->request_body_async_stream, buffer);
+    /* If async-stream, simply call read_to_fill() */
+    if (meta_request->request_body_async_stream != NULL) {
+        return aws_async_input_stream_read_to_fill(meta_request->request_body_async_stream, buffer);
+    }
+
+    /* Else synchronous aws_input_stream */
+    struct aws_input_stream *synchronous_stream =
+        aws_http_message_get_body_stream(meta_request->initial_request_message);
+    AWS_FATAL_ASSERT(synchronous_stream);
+
+    struct aws_future_bool *synchronous_read_future = aws_future_bool_new(meta_request->allocator);
+
+    /* Keep calling read() until we fill the buffer, or hit EOF */
+    struct aws_stream_status status = {.is_end_of_stream = false, .is_valid = true};
+    while ((buffer->len < buffer->capacity) && !status.is_end_of_stream) {
+        /* Read from stream */
+        if (aws_input_stream_read(synchronous_stream, buffer) != AWS_OP_SUCCESS) {
+            aws_future_bool_set_error(synchronous_read_future, aws_last_error());
+            goto synchronous_read_done;
+        }
+
+        /* Check if stream is done */
+        if (aws_input_stream_get_status(synchronous_stream, &status) != AWS_OP_SUCCESS) {
```

Comment:
trivial: do we also need to check `is_valid`?

Reply:
ehhhhhh

if read() fails, or get_status() fails I get an error-code I can report, but if `is_valid` is false I have to make up an error code

we never check `is_valid` in aws-c-http. if it's invalid, it will tell me the next time I try to read()